fs/vfs/fs_timerfd.c

```diff
@@ -0,0 +1,845 @@
+   * counter to zero
+   */
+
+  intflags = enter_critical_section();
+
+  *(FAR timerfd_t *)buffer = dev->counter;
+  dev->counter = 0;
+
+  leave_critical_section(intflags);
+
+  nxsem_post(&dev->exclsem);
+  return sizeof(timerfd_t);
+}
+
+static int timerfd_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+#if 0
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct timerfd_priv_s *priv = inode->i_private;
+#endif
+  return -ENOSYS;
+}
+
+#ifdef CONFIG_TIMER_FD_POLL
+static int timerfd_poll(FAR struct file *filep, FAR struct pollfd *fds,
+                        bool setup)
+{
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct timerfd_priv_s *dev = inode->i_private;
+  int ret;
+  int i;
+
+  ret = nxsem_wait(&dev->exclsem);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  ret = OK;
+
+  if (!setup)
+    {
+      /* This is a request to tear down the poll. */
+
+      FAR struct pollfd **slot = (FAR struct pollfd **)fds->priv;
+
+      /* Remove all memory of the poll setup */
+
+      *slot                = NULL;
+      fds->priv            = NULL;
+      goto errout;
+    }
+
+  /* This is a request to set up the poll. Find an available
+   * slot for the poll structure reference
+   */
+
+  for (i = 0; i < CONFIG_TIMER_FD_NPOLLWAITERS; i++)
+    {
+      /* Find an available slot */
+
+      if (!dev->fds[i])
+        {
+          /* Bind the poll structure and this slot */
+
+          dev->fds[i] = fds;
+          fds->priv   = &dev->fds[i];
+          break;
+        }
+    }
+
+  if (i >= CONFIG_TIMER_FD_NPOLLWAITERS)
+    {
+      fds->priv = NULL;
+      ret       = -EBUSY;
+      goto errout;
+    }
+
+
+  /* Notify the POLLIN event if the counter is not zero */
+
+  if (timerfd_get_counter(dev) > 0)
+    {
+      timerfd_pollnotify(dev, POLLIN);
+    }
+
+errout:
+  nxsem_post(&dev->exclsem);
+  return ret;
+}
+#endif
+
+static void timerfd_timeout_work(FAR void *arg)
+{
+  FAR struct timerfd_priv_s *dev = (FAR struct timerfd_priv_s *)arg;
+  int ret;
+
+  ret = nxsem_wait_uninterruptible(&dev->exclsem);
+  if (ret < 0)
+    {
+      wd_cancel(&dev->wdog);
+      return;
+    }
+
+#ifdef CONFIG_TIMER_FD_POLL
+  /* Notify all poll/select waiters */
+
+  timerfd_pollnotify(dev, POLLIN);
+#endif
+
+  /* Notify all of the waiting readers */
+
+  timerfd_waiter_sem_t *cur_sem = dev->rdsems;
+  while (cur_sem != NULL)
+    {
+      nxsem_post(&cur_sem->sem);
+      cur_sem = cur_sem->next;
+    }
+
+  dev->rdsems = NULL;
+
+  nxsem_post(&dev->exclsem);
+}
+
+static void timerfd_timeout(wdparm_t idev)
+{
+  FAR struct timerfd_priv_s *dev = (FAR struct timerfd_priv_s *)idev;
+  irqstate_t intflags;
+
+  /* Disable interrupts to ensure that expiration counter is accessed
+   * atomically
+   */
+
+  intflags = enter_critical_section();
+
+  /* Increment timer expiration counter.  Do not allow rollover */
+
+  if (dev->counter < TIMER_FD_COUNTER_MAX)
+    {
+      dev->counter++;
```

Comment:
atomic_fetch_add or up_fetchadd32 can avoid the critical secton.

Reply:
I do not want to add dependency to `atomic` for `timerfd` and `up_fetchadd32` is platform specific and dealing with 64 bit incrementing may be a bit tricky.
Do you see a big problem in using of a critical section?

What I see even more, is changing `enter_critical_section`/`leave_critical_section` to `spin_lock_irqsave`/`spin_unlock_irqrestore` to ensure option when `wdog` is running at Core A while non-blocking read happens on Core B. @xiaoxiang781216 what do you think?

Reply:
> I do not want to add dependency to `atomic` for `timerfd` and `up_fetchadd32` is platform specific and dealing with 64 bit incrementing may be a bit tricky. Do you see a big problem in using of a critical section?
> 

No, just a suggestion for your consideration.

> What I see even more, is changing `enter_critical_section`/`leave_critical_section` to `spin_lock_irqsave`/`spin_unlock_irqrestore` to ensure option when `wdog` is running at Core A while non-blocking read happens on Core B. @xiaoxiang781216 what do you think?
> 

Yes, if there is no potential wait in side the critical section, spinlock is better than critical section

> BTW atomic_fetch_add also use `spin_lock_irqsave`/`spin_unlock_irqrestore`

It's depends on the architecture(e.g. arm may use LDREX and STREX).

Reply:
Done